k8s: Crash when k8s cannot init or talk to apiserver

raybejjani opened the pull request:
We want to be more aggressive about indicating errors to the user on
startup. Talking to the k8s server is fairly critical and so exiting on
an error is worthwhile.

Related: #3093

raybejjani committed:
k8s: Crash when k8s cannot init or talk to apiserver

We want to be more aggressive about indicating errors to the user on
startup. Talking to the k8s server is fairly critical and so exiting on
an error is worthwhile.

Signed-off-by: Ray Bejjani <ray@covalent.io>

aanm commented on pkg/k8s/init.go:
```diff
@@ -31,6 +31,9 @@ func Init() error {
 	if err := createDefaultClient(); err != nil {
 		return fmt.Errorf("unable to create k8s client: %s", err)
 	}
+	if _, err := GetServerVersion(); err != nil {
```
Can we add an error if k8s version is less than the minimal version supported by cilium? (`>1.6`)

raybejjani replied:
That is a good idea.  I'll add it.

raybejjani replied:
Done. I opted to bound the range on both ends: `> 1.6, < 1.11`. I'm not sure that this is a good idea, since it means we would break in the future. On the other hand, it explicitly enforces our supported ranges. What do you think?

raybejjani committed:
k8s: Exit if we detect an unsupported k8s version

We wish to indicate incompatibility to a user early. We do this by
returning a fatal error when we we detect an unsupported k8s version.

Signed-off-by: Ray Bejjani <ray@covalent.io>

aanm commented on pkg/k8s/const.go:
```diff
@@ -30,4 +30,9 @@ const (
+	// compatibleK8sVersions is the range of k8s versions this cilium is able to
+	// work with. It will change as we add new support or deprecate older k8s
+	// versions.
+	compatibleK8sVersions = "> 1.6, < 1.11"
```
< 1.12

raybejjani replied:
are we tested to work with 1.11? I wasn't sure how strict to make this. Would it be a "only works with things we tested" or "this will probably work". If the latter, we could just have "> 1.6" since it's simpler.

raybejjani replied:
Chatted offline. I removed the upper bound to avoid forcibly breaking in the future.

tgraf commented on daemon/main.go:
```diff
@@ -721,7 +721,7 @@ func runDaemon() {
 
 	if err := d.EnableK8sWatcher(5 * time.Minute); err != nil {
-		log.WithError(err).Warn("Error while enabling k8s watcher")
+		log.WithError(err).Fatal("Error while enabling k8s watcher")
```
Let's use a less cryptic error message here as this will be quite common:
`"Unable to establish connection to Kubernetes apiserver"`
